### PR TITLE
Add health check gauge

### PR DESCRIPTION
**Pull request by davidxia**

### CODE REVIEW

**Comment by mattnworb** [comment]
I think it would be nicer to have one gauge per healthcheck, rather than one overall gauge, so the values can be reported separately. 

Also I think the same gauges should be added to AgentService as well.

Perhaps something like this in MasterService/AgentService

```java
    // after all environment.healthchecks().register(..) calls...
    for (final String name : environment.healthChecks().getNames()) {
      environment.metrics().register("helios." + name + ".ok", new Gauge<Boolean>() {
        @Override
        public Boolean getValue() {
          final HealthCheck.Result result = environment.healthChecks().runHealthCheck(name);
          return result.isHealthy();
        }
      });
    }

```

**Comment by gimaker** [comment]
For our particular use case, we need the overall metric as well, right? Since we do alerting on aggregate value.

**Comment by mattnworb** [comment]
We should be able to set up an alert if *any* of the healthchecks go to 0; basically a group by `host, healthcheckname`

**Comment by davidxia** [comment]
@spotify/helios-team Finally got the tests to pass.

File: helios-services/pom.xml [hunk 1]
```diff
@@ -554,6 +554,14 @@
...
         </executions>
       </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <source>1.8</source>
+          <target>1.8</target>
+        </configuration>
+      </plugin>
```
Comment (mattnworb) [inline helios-services/pom.xml]: How come this is needed? The same config is in `<pluginManagement>` of the parent pom

File: circle.yml [hunk 2]
```diff
@@ -26,6 +26,8 @@ machine:
   # tells CircleCI that we want virtualenv so that we can pip install codecov without errors
   python:
     version: 2.7.10
+  java:
+    version: oraclejdk8
```
Comment (mattnworb) [inline circle.yml]: looks like this duplicates what is in line 15

**Comment by mattnworb** [comment]
:+1: for the actual healthcheck change but I think the two things pointed out above are not necessary

**Comment by davidxia** [comment]
Thanks. Yea I forgot about those.

**Comment by davidxia** [comment]
@negz @gimaker

File: solo/base/version.txt [hunk 3]
```diff
@@ -1 +1 @@
-0.3
+0.4
```
Comment (gimaker) [inline solo/base/version.txt]: why was this needed? otherwise lgtm

**Comment by gimaker** [comment]
Tests would be nice, too.

File: solo/base/version.txt (hunk 3 above)
Comment (davidxia) [inline solo/base/version.txt]: 0.4 has java8 installed. 0.3 doesn't

**Comment by mattnworb** [comment]
I upgraded java in helios-solo-base back in that original PR, in 0.3

```
$> docker run -it spotify/helios-solo-base:0.3 bin/bash
root@3024bb6592c7:/# java -version
java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)
root@3024bb6592c7:/#
```

but it doesn't hurt to bump the version again

**Comment by davidxia** [comment]
weird, could've sworn I logged into circleCI's heios-solo container after tests failed and ran `java -version` only to see 1.7. /shrugs

File: helios-system-tests/src/main/java/com/spotify/helios/system/AgentArgsBuilder.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    this.httpEndpoint = httpEndpoint;
+    return this;
+  }
+
+  public AgentArgsBuilder adminEndpoint(final String adminEndpoint) {
+    this.adminEndpoint = adminEndpoint;
+    return this;
+  }
+
+  public List<String> build() {
```
Comment (davidxia) [inline helios-system-tests/src/main/java/com/spotify/helios/system/AgentArgsBuilder.java]: Should I add some invariants here like `checkNotNull(dockerHost)`? And if so which ones?

File: helios-system-tests/src/main/java/com/spotify/helios/system/AgentArgsBuilder.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    final ImmutableList.Builder<String> b = ImmutableList.builder();
+    if (verbosity > 0) {
```
Comment (davidxia) [inline helios-system-tests/src/main/java/com/spotify/helios/system/AgentArgsBuilder.java]: These `if`s seem ugly. I'm wondering if there's a prettier way.

**Comment by mattnworb** [comment]
I think a much simpler test would be to extract the anonymous class for the "Gauge of a Healthcheck" to it's own class and write a unit test for that, to see what value it returns when the `Healthcheck` returns different values. Then you wouldn't need these descriptors or pojos or tests that make http calls to a servlet.

File: helios-system-tests/src/main/java/com/spotify/helios/system/SystemTestBase.java [hunk 6]
```diff
@@ -602,7 +615,34 @@ protected AgentMain startDefaultAgent(final String host, final String... args)
         dockerPortRange.upperEndpoint()
     );
     argsList.addAll(asList(args));
-    return startAgent(argsList.toArray(new String[argsList.size()]));
+    return startAgentWithArgs(argsList);
+  }
+
+  protected AgentMain startAgentWithArgs(final List<String> args) throws Exception {
+    final AgentMain agentMain = startAgent(args.toArray(new String[args.size()]));
+    waitForAgentToBeFullyUp();
```
Comment (davidxia) [inline helios-system-tests/src/main/java/com/spotify/helios/system/SystemTestBase.java]: I think this will fix test failures like DeregisterTest.testJobsArePreservedWhenReregistering()` here https://circleci.com/gh/spotify/helios/3795.

**Comment by davidxia** [comment]
@mattnworb I've added a unit test in the last commit. I left the system tests since they test the master and agent are actually registering the gauges and emitting those metrics. Let me know if these tests are not worth it though.

File: helios-services/src/main/java/com/spotify/helios/master/MasterService.java [hunk 7]
```diff
@@ -186,6 +187,12 @@ public MasterService(final MasterConfig config,
 
     environment.lifecycle().manage(zooKeeperHealthChecker);
     environment.healthChecks().register("zookeeper", zooKeeperHealthChecker);
+
+    // Report health checks as a boolean gauge metric
```
Comment (mattnworb) [inline helios-services/src/main/java/com/spotify/helios/master/MasterService.java]: strictly speaking the "boolean" part of these comments are outdated now

Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/master/MasterService.java]: Ah thanks. For context, I changed it to an Integer because every other gauge is numeric, and I didn't know how to make jackson parse something that could be more than one type.

File: helios-system-tests/src/main/java/com/spotify/helios/system/SystemTestBase.java [hunk 8]
```diff
@@ -602,7 +615,34 @@ protected AgentMain startDefaultAgent(final String host, final String... args)
...
+    return agentMain;
+  }
+
+  protected void waitForAgentToBeFullyUp() throws Exception {
```
Comment (mattnworb) [inline helios-system-tests/src/main/java/com/spotify/helios/system/SystemTestBase.java]: this looks identical to `waitForMasterToBeFullUp()`, can you extract the callable so that both functions can use the same code?

Comment (davidxia) [inline helios-system-tests/src/main/java/com/spotify/helios/system/SystemTestBase.java]: Sure thing. I might abandon this as it seems to break a bunch of tests...

**Comment by mattnworb** [comment]
If you think the system test adds value I won't quibble with adding it; although I don't think I would have chosen to test that piece of logic myself.

But:
- adding `Gauge` and `Metrics` as `Descriptor` subclasses feels weird to me, since they really aren't
- Instead of making them Descriptors to gain access to easy JSON to POJO serialization/deserialization, I personally find it easier to just use `ObjectMapper` and it's tree API (JsonNode, ObjectNode) in tests
- I think these classes, and AgentArgsBuilder, should be in src/test/java if they are only for testing purposes and never to be used in the actual main source code

**Comment by davidxia** [comment]
Good points. I'll remove the system test. It added too much code for my
taste anyways.

On Fri, Feb 5, 2016 at 10:49 AM, Matt Brown <notifications@github.com>
wrote:

> If you think the system test adds value I won't quibble with adding it;
> although I don't think I would have chosen to test that piece of logic
> myself.
>
> But:
>
>    - adding Gauge and Metrics as Descriptor subclasses feels weird to me,
>    since they really aren't
>    - Instead of making them Descriptors to gain access to easy JSON to
>    POJO serialization/deserialization, I personally find it easier to just use
>    ObjectMapper and it's tree API (JsonNode, ObjectNode) in tests
>    - I think these classes, and AgentArgsBuilder, should be in
>    src/test/java if they are only for testing purposes and never to be used in
>    the actual main source code
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/helios/pull/840#issuecomment-180411079>.
>



-- 
David Xia
Software Engineer
dxia@spotify.com

**Comment by davidxia** [comment]
finally tests passed

**Comment by mattnworb** [comment]
:+1: :+1:

File: helios-services/src/main/java/com/spotify/helios/agent/AgentService.java [hunk 9]
```diff
@@ -303,9 +305,17 @@ public AgentService(final AgentConfig config, final Environment environment)
     if (!config.getNoHttp()) {
 
       environment.healthChecks().register("docker", dockerHealthChecker);
+      environment.healthChecks().register("zookeeper", zkHealthChecker);
+
+      // Report health checks as a gauge metric
+      environment.healthChecks().getNames().forEach(
+          name -> environment.metrics().register("helios." + name + ".ok", (Gauge<Boolean>) () -> {
```
Comment (gimaker) [inline helios-services/src/main/java/com/spotify/helios/agent/AgentService.java]: `HealthCheckGauge` should probably be used here too.

**Comment by gimaker** [comment]
:+1: aside from my one comment
